File: 1135-connecting-cities-with-minimum-cost/1135-connecting-cities-with-minimum-cost.cpp
class Solution {
private:
    vector<int> parents;
    vector<int> sizes;
    int components;
public:
    int FindRoot(int node){
        if (node == parents[node]) {
            return node;
        }
        return parents[node] = FindRoot(parents[node]);
    }
    void Union(int a, int b) {
        a = FindRoot(a);
        b = FindRoot(b);
        if (a != b) {
            components--;
            if (sizes[a] < sizes[b]) {
				swap(a, b);
            }
            parents[b] = a;
            sizes[a] += sizes[b];
        }
    }
    int kruskal(vector<vector<int>>& connections){
        int minimumcost = 0;
        for(int i = 0; i < connections.size(); i++){
            int u = --connections[i][0];
            int v = --connections[i][1];
            int cost = connections[i][2];
            if (FindRoot(u) != FindRoot(v)){
                minimumcost += cost;
                Union(u, v);
            }
        }
        return components != 1 ? -1 : minimumcost;
    }
    int minimumCost(int n, vector<vector<int>>& connections) {
        parents.resize(n);
        iota(parents.begin(), parents.end(), 0);
        sizes.resize(n, 1);
        components = n;
        sort(connections.begin(), connections.end(), [](auto &a, auto &b) {
            return a[2] < b[2]; 
        });
        int cost = kruskal(connections);
        return cost;
    }
};

File: 1135-connecting-cities-with-minimum-cost/1135-connecting-cities-with-minimum-cost.py
class UnionFind:
    def __init__(self, n):
        self.parent = list(range(n))
        self.size = [1] * n
        self.cities = n

    def find(self, x):
        if x != self.parent[x]:
            self.parent[x] = self.find(self.parent[x]) # Path compression
        return self.parent[x]

    def union(self, a, b):
        x = self.find(a)
        y = self.find(b)
        if x != y:
            self.cities -= 1
            if self.size[x] < self.size[y]:
                x, y = y, x
            self.parent[y] = x
            self.size[x] += self.size[y]
            
class Solution:
    def minimumCost(self, n: int, connections: List[List[int]]) -> int:
        # Sort edges by cost
        connections.sort(key=lambda edge: edge[2])

        # Kruskal's Algorithm
        dsu = UnionFind(n)
        minimum_cost = 0
        for i, j, cost in connections:
            if dsu.find(i - 1) != dsu.find(j - 1):
                minimum_cost += cost
                dsu.union(i - 1, j - 1)

        return minimum_cost if dsu.cities == 1 else -1 
